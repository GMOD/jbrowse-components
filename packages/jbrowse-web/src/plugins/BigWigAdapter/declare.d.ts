declare module 'abortable-promise-cache'
