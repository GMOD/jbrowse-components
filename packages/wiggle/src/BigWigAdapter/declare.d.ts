declare module '@gmod/jbrowse-core/util/QuickLRU'
