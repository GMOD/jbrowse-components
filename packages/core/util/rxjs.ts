import { Observable, Observer } from 'rxjs'
import { takeUntil } from 'rxjs/operators'
import { observeAbortSignal } from '.'

/**
 * Wrapper for rxjs Observable.create with improved error handling and
 * aborting support
 * @param func - observer function, could be async
 */
export function ObservableCreate<T>(
  func: (arg: Observer<T>) => void | Promise<void>,
  signal?: AbortSignal,
): Observable<T> {
  return Observable.create((observer: Observer<T>) => {
    try {
      const ret = func(observer)
      // catch async errors
      if (ret && ret.catch) {
        ret.catch((error: Error) => observer.error(error))
      }
    } catch (error) {
      // catch sync errors
      observer.error(error)
    }
  }).pipe(takeUntil(observeAbortSignal(signal)))
}
