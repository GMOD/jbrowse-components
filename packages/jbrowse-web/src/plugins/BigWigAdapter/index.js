import Plugin from '../../Plugin'
import AdapterType from '../../pluggableElementTypes/AdapterType'
import { ConfigurationSchema } from '../../configuration'
import AdapterClass from './BigWigAdapter'

const configSchema = ConfigurationSchema(
  'BigWigAdapter',
  {
    bigWigLocation: {
      type: 'fileLocation',
      defaultValue: { uri: '/path/to/my.bw' },
    },
    assemblyName: {
      type: 'string',
      defaultValue: '',
    },
  },
  { explicitlyTyped: true },
)

export default class BigWigAdapterPlugin extends Plugin {
  install(pluginManager) {
    pluginManager.addAdapterType(
      () =>
        new AdapterType({
          name: 'BigWigAdapter',
          configSchema,
          AdapterClass,
        }),
    )
  }
}
