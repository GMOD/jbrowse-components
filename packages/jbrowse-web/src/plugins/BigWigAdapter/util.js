/*
 * calculate standard deviation using the 'shortcut method' that accepts
 * the sum and the sum squares of the elements
 *
 * @param sum - sum(i, 1..n)
 * @param sumSquares - sum(i^2, 1..n)
 * @param n - number of elements
 * @param population - boolean: use population instead of sample correction
 * @return the estimated std deviation
 */
export function calcStdFromSums(sum, sumSquares, n, population = false) {
  if (n === 0) return 0
  let variance
  if (population) {
    variance = sumSquares / n - (sum * sum) / (n * n)
  } else {
    // sample correction is n-1
    variance = sumSquares - (sum * sum) / n
    if (n > 1) {
      variance /= n - 1
    }
  }

  return variance < 0 ? 0 : Math.sqrt(variance)
}

/*
 * @param s - a summary stats object with scoreSum, featureCount, scoreSumSquares, and basesCovered
 * @return - a summary stats object with scoreMean, scoreStdDev, and featureDensity added
 */
export function rectifyStats(s) {
  return {
    ...s,
    scoreMean: s.featureCount ? s.scoreSum / s.featureCount : 0,
    scoreStdDev: calcStdFromSums(s.scoreSum, s.scoreSumSquares, s.featureCount),
    featureDensity: s.featureCount / s.basesCovered,
  }
}

/*
 * calculates per-base scores for variable width features over a region
 * @param region - object contains start, end
 * @param features - list of features with start, end, score
 * @return array of numeric scores
 */
export function calcRealStats(region, features) {
  const { start, end } = region
  const scores = []
  const feats = features.sort((a, b) => a.start < b.start)
  let pos = start
  let currentFeat = 0
  let i = 0
  while (pos < end) {
    while (currentFeat < feats.length && pos >= feats[currentFeat].end) {
      currentFeat += 1
    }
    const f = feats[currentFeat]
    // console.log('currentPos', pos, currentFeat)
    if (!f) {
      scores[i] = 0
    } else if (pos >= f.start && pos < f.end) {
      scores[i] = f.score
    } else {
      scores[i] = 0
    }
    i += 1
    pos += 1
  }
  return scores
}

/*
 * transform a list of scores to summary statistics
 * @param region - object with start, end
 * @param feats - array of features which are possibly summary features
 * @return - object with scoreMax, scoreMin, scoreSum, scoreSumSquares, etc
 */
export function scoresToStats(region, feats) {
  const { start, end } = region
  let scoreMax = -Infinity
  let scoreMin = Infinity
  let scoreSum = 0
  let scoreSumSquares = 0

  for (let i = 0; i < feats.length; i += 1) {
    const f = feats[i]
    scoreMax = Math.max(scoreMax, f.summary ? f.maxScore : f.score)
    scoreMin = Math.min(scoreMin, f.summary ? f.minScore : f.score)
    scoreSum += f.score
    scoreSumSquares += f.score * f.score
  }

  return rectifyStats({
    scoreMax,
    scoreMin,
    scoreSum,
    scoreSumSquares,
    featureCount: feats.length,
    basesCovered: end - start + 1,
  })
}
