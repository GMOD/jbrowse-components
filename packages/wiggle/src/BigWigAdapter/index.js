import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'

export { default as AdapterClass } from './BigWigAdapter'

export const configSchema = ConfigurationSchema(
  'BigWigAdapter',
  {
    bigWigLocation: {
      type: 'fileLocation',
      defaultValue: { uri: '/path/to/my.bw' },
    },
  },
  { explicitlyTyped: true },
)
