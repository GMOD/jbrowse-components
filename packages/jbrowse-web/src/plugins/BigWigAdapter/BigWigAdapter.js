import { BigWig } from '@gmod/bbi'
import BaseAdapter from '../../BaseAdapter'
import { openLocation } from '../../util/io'
import SimpleFeature from '../../util/simpleFeature'
import { ObservableCreate } from '../../util/rxjs'

function blankStats(covered) {
  return {
    scoreMax: -Infinity,
    scoreMin: Infinity,
    scoreSum: 0,
    scoreSumSquares: 0,
    basesCovered: covered,
    featureCount: 0,
  }
}

function calcStdFromSums(sum, sumSquares, n) {
  if (n === 0) return 0

  let variance = sumSquares - (sum * sum) / n
  if (n > 1) {
    variance /= n - 1
  }
  return variance < 0 ? 0 : Math.sqrt(variance)
}

function rectifyStats(s) {
  return {
    ...s,
    scoreMean: s.featureCount ? s.scoreSum / s.featureCount : 0,
    scoreStdDev: calcStdFromSums(s.scoreSum, s.scoreSumSquares, s.featureCount),
    featureDensity: s.featureCount / s.basesCovered,
  }
}

export default class BigWigAdapter extends BaseAdapter {
  static capabilities = ['getFeatures', 'getRefNames']

  constructor(config) {
    super(config)
    this.bigwig = new BigWig({
      filehandle: openLocation(config.bigWigLocation),
    })
  }

  async getRefNames() {
    const header = await this.bigwig.getHeader()
    return Object.keys(header.refsByName)
  }

  async refIdToName(refId) {
    return ((await this.bigwig.getHeader()).refsByNumber[refId] || {}).name
  }

  async getGlobalStats() {
    const header = await this.bigwig.getHeader()
    return header.totalSummary
  }

  // todo: add caching
  // todo: incorporate summary blocks
  async getRegionStats({ refName, start, end }) {
    console.time('getf')
    const feats = await this.bigwig.getFeatures(refName, start, end)
    console.timeEnd('getf')
    console.log(refName, start, end)
    const blank = blankStats(end - start)
    if (!feats.length) {
      blank.scoreMax = 0
      blank.scoreMin = 0
    }
    const stats = feats.reduce((a, b) => {
      const s = b.score || 0
      return {
        ...a,
        scoreMax: Math.max(a.scoreMax, s),
        scoreMin: Math.min(a.scoreMin, s),
        scoreSum: a.scoreSum + s,
        scoreSumSquares: a.scoreSumSquares + s * s,
        featureCount: a.featureCount + 1,
      }
    }, blank)
    return rectifyStats(stats)
  }

  // todo: add caching
  async getMultiRegionStats(regions = []) {
    if (!regions.length) return undefined
    const ret = await Promise.all(regions.map(r => this.getRegionStats(r)))
    const s = ret.reduce((a, b) => {
      return {
        ...a,
        scoreMax: Math.max(a.scoreMax, b.scoreMax),
        scoreMin: Math.min(a.scoreMin, b.scoreMin),
        scoreSum: a.scoreSum + b.scoreSum,
        scoreSumSquares: a.scoreSumSquares + b.scoreSumSquares,
        featureCount: a.featureCount + b.featureCount,
        basesCovered: a.basesCovered + b.basesCovered,
      }
    }, blankStats(0))
    return rectifyStats(s)
  }

  /**
   * Fetch features for a certain region
   * @param {Region} param
   * @returns {Observable[Feature]} Observable of Feature objects in the region
   */

  getFeatures({ /* assembly, */ refName, start, end }, signal) {
    return ObservableCreate(async observer => {
      const ob2 = await this.bigwig.getFeatureStream(refName, start, end, {
        signal,
      })
      ob2.subscribe(
        chunk => {
          chunk.forEach(record => {
            observer.next(
              new SimpleFeature({
                id: record.start + 1,
                data: record,
              }),
            )
          })
        },
        error => {
          throw new Error(error)
        },
        () => observer.complete(),
      )
    })
  }
}
