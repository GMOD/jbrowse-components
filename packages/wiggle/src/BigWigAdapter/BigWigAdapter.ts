import { BigWig, Feature as BBIFeature } from '@gmod/bbi'
import {
  BaseFeatureDataAdapter,
  BaseOptions,
} from '@gmod/jbrowse-core/data_adapters/BaseAdapter'
import { NoAssemblyRegion } from '@gmod/jbrowse-core/util/types'
import { openLocation } from '@gmod/jbrowse-core/util/io'
import { ObservableCreate } from '@gmod/jbrowse-core/util/rxjs'
import SimpleFeature, { Feature } from '@gmod/jbrowse-core/util/simpleFeature'
import { map, mergeAll } from 'rxjs/operators'
import { readConfObject } from '@gmod/jbrowse-core/configuration'
import { Instance } from 'mobx-state-tree'
import {
  blankStats,
  rectifyStats,
  scoresToStats,
  UnrectifiedFeatureStats,
  DataAdapterWithGlobalStats,
} from '../statsUtil'

import configSchema from './configSchema'

export default class BigWigAdapter extends BaseFeatureDataAdapter
  implements DataAdapterWithGlobalStats {
  private bigwig: BigWig

  public constructor(config: Instance<typeof configSchema>) {
    super(config)
    this.bigwig = new BigWig({
      filehandle: openLocation(readConfObject(config, 'bigWigLocation')),
    })
  }

  public async getRefNames(opts: BaseOptions = {}) {
    const header = await this.bigwig.getHeader(opts)
    return Object.keys(header.refsByName)
  }

  public async refIdToName(refId: number) {
    const h = await this.bigwig.getHeader()
    return (h.refsByNumber[refId] || { name: undefined }).name
  }

  public async getGlobalStats(opts: BaseOptions = {}) {
    const header = await this.bigwig.getHeader(opts)
    return rectifyStats(header.totalSummary as UnrectifiedFeatureStats)
  }

  // todo: incorporate summary blocks
  public getRegionStats(region: NoAssemblyRegion, opts: BaseOptions = {}) {
    const feats = this.getFeatures(region, opts)
    return scoresToStats(region, feats)
  }

  // todo: add caching
  public async getMultiRegionStats(
    regions: NoAssemblyRegion[] = [],
    opts: BaseOptions = {},
  ) {
    if (!regions.length) {
      return blankStats()
    }
    const feats = await Promise.all(
      regions.map(region => this.getRegionStats(region, opts)),
    )

    const scoreMax = feats
      .map(s => s.scoreMax)
      .reduce((acc, curr) => Math.max(acc, curr))
    const scoreMin = feats
      .map(s => s.scoreMin)
      .reduce((acc, curr) => Math.min(acc, curr))
    const scoreSum = feats.map(s => s.scoreSum).reduce((a, b) => a + b, 0)
    const scoreSumSquares = feats
      .map(s => s.scoreSumSquares)
      .reduce((a, b) => a + b, 0)
    const featureCount = feats
      .map(s => s.featureCount)
      .reduce((a, b) => a + b, 0)
    const basesCovered = feats
      .map(s => s.basesCovered)
      .reduce((a, b) => a + b, 0)

    return rectifyStats({
      scoreMin,
      scoreMax,
      featureCount,
      basesCovered,
      scoreSumSquares,
      scoreSum,
    })
  }

  public getFeatures(region: NoAssemblyRegion, opts: BaseOptions = {}) {
    const { refName, start, end } = region
    const { signal, bpPerPx, headers } = opts
    return ObservableCreate<Feature>(async observer => {
      const ob = await this.bigwig.getFeatureStream(refName, start, end, {
        signal,
        basesPerSpan: bpPerPx,
        headers,
      })
      ob.pipe(
        mergeAll(),
        map((record: BBIFeature) => {
          return new SimpleFeature({
            id: String(record.start + 1),
            data: record,
          })
        }),
      ).subscribe(observer)
    }, opts.signal)
  }

  public freeResources(): void {}
}
