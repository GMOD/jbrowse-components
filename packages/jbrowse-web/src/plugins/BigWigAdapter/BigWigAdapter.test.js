import { toArray } from 'rxjs/operators'

import BigWigAdapter from './BigWigAdapter'

test('adapter can fetch features from volvox.bw', async () => {
  const adapter = new BigWigAdapter({
    assemblyName: 'volvox',
    bigWigLocation: { path: require.resolve('./test_data/volvox.bw') },
  })

  const features = adapter.getFeaturesInRegion({
    assembly: 'volvox',
    refName: 'ctgA',
    start: 0,
    end: 20000,
  })

  const featuresArray = await features.pipe(toArray()).toPromise()
  const featuresJsonArray = featuresArray.map(f => f.toJSON())
  expect(featuresJsonArray.slice(1000, 1010)).toMatchSnapshot()
})
