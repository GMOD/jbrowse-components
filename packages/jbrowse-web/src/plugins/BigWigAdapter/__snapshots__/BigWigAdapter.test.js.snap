// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adapter can fetch features from volvox.bw 1`] = `
Array [
  Object {
    "end": 1003,
    "parentID": undefined,
    "score": 18,
    "start": 1002,
  },
  Object {
    "end": 1004,
    "parentID": undefined,
    "score": 18,
    "start": 1003,
  },
  Object {
    "end": 1005,
    "parentID": undefined,
    "score": 18,
    "start": 1004,
  },
  Object {
    "end": 1006,
    "parentID": undefined,
    "score": 19,
    "start": 1005,
  },
  Object {
    "end": 1007,
    "parentID": undefined,
    "score": 20,
    "start": 1006,
  },
  Object {
    "end": 1008,
    "parentID": undefined,
    "score": 20,
    "start": 1007,
  },
  Object {
    "end": 1009,
    "parentID": undefined,
    "score": 20,
    "start": 1008,
  },
  Object {
    "end": 1010,
    "parentID": undefined,
    "score": 20,
    "start": 1009,
  },
  Object {
    "end": 1011,
    "parentID": undefined,
    "score": 20,
    "start": 1010,
  },
  Object {
    "end": 1012,
    "parentID": undefined,
    "score": 19,
    "start": 1011,
  },
]
`;
